danceradvent/public/articles/2015/7-medium-scale-part-5.pod

```diff
@@ -0,0 +1,208 @@
+defines C<POST> and C<PUT> so that they're nearly interchangeable, which
+is how we ended up with competing REST/CRUD API design philosophies.
+
+Some web API designers like to use C<POST> for both create and update
+operations since you can distinguish the cases based on whether the
+HTTP request contains a database row ID: if so, it's a request to update
+an existing record, else it's a record creation request.
+
+That justification is fine in a web framework that defines the URL
+scheme in the file system, such as PHP, ASP, or JSP since it is normal
+to handle all four verbs in a single file in that sort of web framework.
+In Dancer, my preferred scheme works better since it defines one route
+per CRUD operation.
+
+Although the HTTP spec defines C<PUT> and C<POST> as partially
+interchangeable, you must not swap the C<PUT> and C<POST> verbs from the
+way I've defined them above. The HTTP spec says
+L<C<PUT> is idempotent|http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.1>, meaning
+that the browser is allowed to cache C<PUT> requests, suppressing
+subsequent identical calls if the request parameters do not change. In
+CRUD terms, this means that if you define "create" in terms of C<PUT>,
+the browser might not let you create two records with identical
+contents but different IDs, which means you risk losing data. The HTTP
+spec does not allow browsers to assume that identical C<POST> calls are
+idempotent, however, so we use that verb for CRUD's create operation,
+leaving C<PUT> = update.
+
+That's enough design philosophy; how does all that affect our Dancer
+code? We might decide that since we've almost got a REST API here that
+we might want to push it the rest of the way.
+
+Let's move the CRUD-like route handlers in C<lib/App/MajorFeature.pm> to
+C<lib/App/API/SubFeature.pm>, and also refine the programming interface
+a bit:
+
+ prefix '/api' => sub {
+     prefix '/subfeature' => sub {
+         post '/'    => \&_post;
+         get  '/'    => \&_get;
+         put  '/:id' => \&_put;
+         del  '/:id' => \&_del;
+     };
+ };
+
+My choice to use leading underscores in these module-internal route
+handler function names lets us parallel the Dancer DSL keyword naming
+scheme.
+
+You might instead prefer to make the HTTP-to-CRUD mapping more explicit:
+
+ prefix '/api' => sub {
+     prefix '/subfeature' => sub {
+         post '/'    => \&_create;
+         get  '/'    => \&_read;
+         put  '/:id' => \&_update;
+         del  '/:id' => \&_delete;
+     };
+ };
+
+As in the previous article in this series, the leading underscore
+convention saves us some aggravation here since C<delete> is a Perl
+keyword, and C<read> is a core Perl function.
+
+The only other thing to change here is that REST APIs normally return
+JSON or XML. I prefer JSON since that's directly usable by the
+JavaScript code that made the Ajax call to one of these APIs. One of the
+best features of Dancer is that automatic JSON encoding is built right
```

Comment:
You will need to set the serializer to JSON, either in your config:

    # config.yml
    serializer: JSON

Or in your application:

    set 'serializer' => 'JSON';